Modules/posixmodule.c

```diff
@@ -14962,6 +14962,318 @@ os_waitstatus_to_exitcode_impl(PyObject *module, PyObject *status_obj)
 #endif
 
 
+#if defined(MS_WINDOWS)
+/*[clinic input]
+os._isdir
+
+    path: 'O'
+
+Return True if path is an existing directory.
+
+This follows symbolic links, so both islink() and isdir() can be true for the
+same path.
+
+[clinic start generated code]*/
+
+static PyObject *
+os__isdir_impl(PyObject *module, PyObject *path)
+/*[clinic end generated code: output=e87264f004b1aa44 input=2b77ce7f601eb6bc]*/
+{
+    HANDLE hfile;
+    BOOL close_file = TRUE;
+    FILE_BASIC_INFO info = { 0 };
+    path_t _path = PATH_T_INITIALIZE("isdir", "path", 0, 1);
+    STRUCT_STAT st;
+    DWORD error;
+    int result;
+
+    if (!path_converter(path, &_path)) {
+        path_cleanup(&_path);
+        if (PyErr_ExceptionMatches(PyExc_ValueError)) {
+            PyErr_Clear();
+            Py_RETURN_FALSE;
+        } else {
+            return NULL;
+        }
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+    if (_path.fd != -1) {
+        hfile = _Py_get_osfhandle_noraise(_path.fd);
+        close_file = FALSE;
+    } else {
+        hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
+                            OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    }
+    if (hfile != INVALID_HANDLE_VALUE) {
+        GetFileInformationByHandleEx(hfile, FileBasicInfo, &info, sizeof(info));
```

Comment:
There's a failure condition here that may need handling (maybe not a _real_ one, since we've got the handle already, but a potential one).

Reply:
All `OSError` exceptions are ignored by `os.path.isdir()`, etc. If `GetFileInformationByHandleEx()` fails, then `FILE_ATTRIBUTE_DIRECTORY` won't be set in `info.FileAttributes`. 

Alternatively, we could skip initializing `info` to zeros, and only check `info` if `GetFileInformationByHandleEx()` succeeds.